pkg/config/provider.go

```diff
@@ -1,29 +1,230 @@
+	// the generated resources, one can add "aws_shield_protection_group$".
+	// To include whole aws waf group, one can add "aws_waf.*" to the list.
+	// Defaults to []string{".+"} which would include all resources.
+	IncludeList []string
+
+	// Resources is a map holding resource configurations where key is Terraform
+	// resource name.
+	Resources map[string]*Resource
+
+	// resourceConfigurators is a map holding resource configurators where key
+	// is Terraform resource name.
+	resourceConfigurators map[string]ResourceConfiguratorChain
+}
+
+// A ProviderOption configures a Provider.
+type ProviderOption func(*Provider)
+
+// WithGroupSuffix configures GroupSuffix for resources of this Provider.
+func WithGroupSuffix(s string) ProviderOption {
+	return func(p *Provider) {
+		p.GroupSuffix = s
+	}
+}
+
+// WithShortName configures ShortName for resources of this Provider.
+func WithShortName(s string) ProviderOption {
+	return func(p *Provider) {
+		p.ShortName = s
+	}
+}
+
+// WithIncludeList configures IncludeList for this Provider.
+func WithIncludeList(l []string) ProviderOption {
+	return func(p *Provider) {
+		p.IncludeList = l
+	}
 }
 
-// SetForResource sets configuration for a given resource.
-func (p *Provider) SetForResource(resource string, cfg Resource) {
-	defer p.mx.Unlock()
-	p.mx.Lock()
-	p.resource[resource] = cfg
+// WithSkipList configures SkipList for this Provider.
+func WithSkipList(l []string) ProviderOption {
+	return func(p *Provider) {
+		p.SkipList = l
+	}
+}
+
+// WithXPProviderConfig configures XPProviderConfig for this Provider.
+func WithXPProviderConfig(c XPProviderConfig) ProviderOption {
+	return func(p *Provider) {
+		p.XPProviderConfig = c
+	}
+}
+
+// WithDefaultResourceFn configures DefaultResourceFn for this Provider
+func WithDefaultResourceFn(f DefaultResourceFn) ProviderOption {
+	return func(p *Provider) {
+		p.DefaultResourceFn = f
+	}
+}
+
+// NewProvider builds and returns a new Provider.
+func NewProvider(resourceMap map[string]*schema.Resource, prefix string, modulePath string, opts ...ProviderOption) Provider {
+	p := Provider{
+		ModulePath:              modulePath,
+		TerraformResourcePrefix: fmt.Sprintf("%s_", prefix),
+		GroupSuffix:             fmt.Sprintf(".%s.tf.crossplane.io", prefix),
+		ShortName:               fmt.Sprintf("tf%s", prefix),
+		DefaultResourceFn:       getDefaultResource,
+		XPProviderConfig: XPProviderConfig{
+			APIVersion:        defaultAPIVersion,
+			ControllerPackage: "providerconfig",
+		},
+
+		IncludeList: []string{
+			// Include all Resources
+			".+",
+		},
+		Resources: map[string]*Resource{},
+
+		resourceConfigurators: map[string]ResourceConfiguratorChain{},
+	}
+
+	for _, o := range opts {
+		o(&p)
+	}
+
+	p.parseSchema(resourceMap)
+
+	return p
+}
+
+// AddResourceConfigurator adds resource specific configurators.
+func (p *Provider) AddResourceConfigurator(resource string, c ResourceConfiguratorFn) { //nolint:interfacer
```

Comment:
I can imagine that we can already achieve that by adding a resource-specific configurator (i.e. in config/group/config.go) which overrides whatever we want. Or am I missing something, could you elaborate a bit?

Reply:
I'm talking about overriding all of the configurators, something like `regardless of what's added to the chain, i want to use only this specific resource configurator function`. Even though we're able to access `Resources` field, we can't really remove a configurator or reset that func array to 0 for a certain resource since only `AddResourceConfigurator` is available.

Reply:
Resource-specific configurators are added at the end, so whatever configurator is added there overrides all the previous ones since it will run as the last one. However, I think I see your point, instead of having to override all fields to make sure you have the right configuration, just run this configurator on top of default. 
I'll add `SetResourceConfigurator`.